pkg/agent/nodeportlocal/k8s/controller.go

```diff
@@ -112,12 +173,122 @@ func (c *Controller) EnqueueObj(obj interface{}) {
 	c.queue.Add(podKey)
 }
 
-func (c *Controller) Worker() {
+func (c *NPLController) checkDeletedSvc(obj interface{}) (*corev1.Service, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("Received unexpected object: %v", obj)
+	}
+	svc, ok := deletedState.Obj.(*corev1.Service)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Service: %v", deletedState.Obj)
+	}
+	return svc, nil
+}
+
+func (c *NPLController) enqueueSvcUpdate(oldObj, newObj interface{}) {
+	// In case where the app selector in Service gets updated from one valid selector to another
+	// both sets of Pods (corresponding to old and new selector) need to be considered
+
+	// Donot push to queue if 1) Service ResourceVersions don't change OR
+	// 2) Service spec selectors AND 3) NPLEnabledAnnotationKey values don't change
+	newSvc := newObj.(*corev1.Service)
+	oldSvc := oldObj.(*corev1.Service)
+	if oldSvc.ResourceVersion == newSvc.ResourceVersion ||
+		reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector) &&
+			oldSvc.Annotations[NPLEnabledAnnotationKey] == newSvc.Annotations[NPLEnabledAnnotationKey] {
+		return
+	}
+
+	// disjunctive union of Pods from both Service sets
+	oldPodSet := sets.NewString(c.getPodsFromService(oldSvc)...)
+	newPodSet := sets.NewString(c.getPodsFromService(newSvc)...)
+	podKeys := oldPodSet.Difference(newPodSet).Union(newPodSet.Difference(oldPodSet))
```

Comment:
Does it mean we do not enqueue the common set of old and new?

Reply:
yes, if a pod is present in both the oldSvc and the newSvc sets, then we don't need to enqueue that.
Consider the sets of Service - {Pods}
s1 - {1,2,3} and s2 - {3,4,5}
then we only need to queue {1,2,4,5}, since queueing 3 won't change anything on pod 3.

Reply:
How if Service's NPL annotation changes?

Reply:
same as replied above, if NPL annotation change from valid to invalid, the valid one will lead to a set of pods, while the invalid one won't, because of the `getPodsFromService` function. The function only returns a slice of pods only if the service has the valid NPL annotation.

Reply:
Got it now. Thanks for the explanation.